hypothesis-python/src/hypothesis/extra/array_api.py

```diff
@@ -894,6 +1063,25 @@ def __repr__(self):
         indices=indices,
     )
 
+    if api_version > "2021.12":
+
+        @defines_strategy()
+        def complex_dtypes(
+            *, sizes: Union[int, Sequence[int]] = (64, 128)
+        ) -> st.SearchStrategy[DataType]:
+            return _complex_dtypes(xp, sizes=sizes)
+
+        complex_dtypes.__doc__ = _complex_dtypes.__doc__
+        kwargs["complex_dtypes"] = complex_dtypes
+
+    namespace = StrategiesNamespace(**kwargs)
+    try:
+        _args_to_xps[(xp, api_version)] = namespace
```

Comment:
We want to cache the `api_version=None` case too, so let's add:
```python
        if inferred_version:
            _args_to_xps[(xp, None)] = namespace
```
plus corresponding test cases to check both that None is cached, and that passing the inferred version hits the newly-cached version (and vice-versa).

Reply:
Good shout! All happy on my end if you want to merge after checking the commit today that implements this :slightly_smiling_face: